feat: Set current namespace for in-cluster SDK in multi-user mode and add healthz endpoint to API backends

DavidSpek opened the pull request:
**Description of your changes:**
When running KFP in notebooks in a multi-user environment is in handy to have it automatically set the current namespace in the local KFP context. I first [added similar ](https://github.com/kubeflow-kale/kale/pull/219)code to Kale, however, it was advised that this should be done on the KFP SDK side.

Bobgy commented:
This was previously attempted at https://github.com/kubeflow/pipelines/pull/4293, but we found it's a breaking change.

elikatsis commented:
Assuming what namespace to set is a breaking change, indeed.
What we could keep from this PR is the creation of the config directory. Because if it is not, setting the namespaces raises an error. This error needs to be handled by the KFP SDK.

DavidSpek commented:
@Bobgy I think the implementation is quite a bit different from https://github.com/kubeflow/pipelines/pull/4293, as that changed the `_load_context_settings_or_default` function which would load the context setting from the file and if that did not exist grab it from the pod. With this PR it will only grab the namespace if no namespace has been set yet and only if it detects it is running from a jupyter notebook. I think the main problem with https://github.com/kubeflow/pipelines/pull/4293/files is the fact that it changes the `_load_context_settings_or_default` behavior for all scenarios. I'd be interested to see what the tests show, and if needed I'll try to fix it or find a different/better solution.

Bobgy commented:
/ok-to-test
Sure, let's run the tests first

DavidSpek commented:
@Bobgy Am I correct in concluding the sample-test uses notebooks for the sample-test pods? Is there a method that is used to detect if it is running in single or multi-user mode?

DavidSpek commented:
After looking into the error I am wondering why the namespace cannot be passed to the backend in single user mode. The experiment_server.go file mentions "In single user mode, apply filter with empty namespace for backward compatibile". As all pipelines run in a namespace the parameter should be passable in single user mode, especially as the KFP SDK seems to use namespace `kubeflow` by default. I may not be the best judge of this, but I feel like many of the Kubeflow deployments are or will be multi-user ones and given the incorporation of notebooks in Kubeflow I think setting the KFP namespace automatically from within a notebook is a good feature to have.

Would it be possible to allow (but not require) the passing of a namespace in single user mode as an alternative? If that is not an option, how would one go about adding [IsMultiUserMode](https://github.com/kubeflow/pipelines/blob/d639b80043998438e1ffc8525e708431c46d2b35/backend/src/apiserver/common/config.go#L80) to the API so that the KFP SDK can properly check what environment it is running in?

DavidSpek commented:
@Bobgy @elikatsis I have added a `get_kfp_healthz` function that returns the API response from `/apis/v1beta1/healthz` as json. I have also added `multi_user` to `/apis/v1beta1/healthz` to be able to check if KFP is running in a single or multi user environment.

Bobgy commented on sdk/python/kfp/_client.py:
```diff
@@ -277,15 +280,35 @@ def set_user_namespace(self, namespace):
       namespace: kubernetes namespace from the local context file or empty if it wasn't set.
     """
+    if self._context_setting['namespace'] == "":
+      if self.get_kfp_healthz()['multi_user'] is True:
```
Suggest moving the check to Client instantiation, so that it only needs to be called once. Multi user mode should be a constant.

elikatsis replied:
+1

Also, maybe do `self.get_kfp_healthz.get("multi_user")` so that it does not break if an old apiserver image is running

Bobgy reviewed (commented):
Thank you! Mostly looking good

elikatsis commented:
@DavidSpek this is great! Thank you

As long as you address Yuan's last comment, it looks good to me

DavidSpek commented:
@Bobgy @elikatsis Thanks for the suggestions, I believe I have implemented them correctly now. I was wondering if you have comments on my addition to the `main.go` file. Having 0 experience with Go and a total of 1 month of developing experience from working on Kale I'm not sure if the `IsMultiUserMode` will correctly be formatted as a boolean in the eventual JSON output. Without `strconv.FormatBool` I was getting a mismatched types error and I believe due to the formatting of the total string the result in the JSON will be a boolean. However, I'd like to get your opinion(s) on it given that I don't have an easy way of testing this change locally.

Bobgy commented on backend/src/apiserver/main.go:
```diff
@@ -140,7 +141,7 @@ func startHttpProxy(resourceManager *resource.ResourceManager) {
 	topMux.HandleFunc("/apis/v1beta1/pipelines/upload_version", pipelineUploadServer.UploadPipelineVersion)
 	topMux.HandleFunc("/apis/v1beta1/healthz", func(w http.ResponseWriter, r *http.Request) {
-		io.WriteString(w, `{"commit_sha":"`+common.GetStringConfigWithDefault("COMMIT_SHA", "unknown")+`", "tag_name":"`+common.GetStringConfigWithDefault("TAG_NAME", "unknown")+`"}`)
+		io.WriteString(w, `{"commit_sha":"`+common.GetStringConfigWithDefault("COMMIT_SHA", "unknown")+`", "tag_name":"`+common.GetStringConfigWithDefault("TAG_NAME", "unknown")+`", "multi_user":`+strconv.FormatBool(common.IsMultiUserMode())+`}`)
```
Hi, I just realized there's some complexities here. KFP SDK usually connect to KFP UI server in fact, KFP UI server redirects all api requests to KFP API Server, but the healthz endpoint is not redirected.

So you need to make sure both healthz endpoint expose the multi user option.
You can find UI server's healthz endpoint at https://github.com/kubeflow/pipelines/blob/80e1d7063dc6eea0d17606c26fb61aff5eca4310/frontend/server/handlers/healthz.ts#L81.

DavidSpek replied:
Thanks for catching that, I wouldn't have found that by myself. I hope this is addressed properly in my latest commit.

Bobgy replied:
Thank you @DavidSpek!

First, because right now we support people using KFP SDK with KFP UI/KFP API endpoint interchangeably, so I'd suggest using the same field name for `multi_user` field, you can choose to expose both `apiServerMultiUser` and `multi_user` fields in UI server as a stopgap solution.

Second, would you mind adding an integration test for the `healthz` endpoint for UI server?
You can find existing tests in https://github.com/kubeflow/pipelines/blob/898dff37320d128de393a0146d8b7f1009aff55a/frontend/server/app.test.ts#L115.

Bobgy commented on sdk/python/kfp/_client.py:
```diff
@@ -277,9 +285,22 @@ def set_user_namespace(self, namespace):
+    Returns:
+      response: json formatted response from the healtz endpoint.
+    """
+    healthz_api = 'http://' + Client.IN_CLUSTER_DNS_NAME.format('kubeflow') + '/' + Client.HEALTH_PATH
```
We have `host` passed in as an optional parameter. We should use it or the resolved host instead of always using IN_CLUSTER_DNS_NAME

DavidSpek replied:
@Bobgy I have changed it to `self._existing_config.host`, which I believe is the correct to grab the resolved host.

Bobgy commented on sdk/python/kfp/_client.py:
```diff
@@ -277,9 +285,22 @@ def set_user_namespace(self, namespace):
+      response: json formatted response from the healtz endpoint.
+    """
+    healthz_api = 'http://' + Client.IN_CLUSTER_DNS_NAME.format('kubeflow') + '/' + Client.HEALTH_PATH
+    r = requests.get(healthz_api)
```
shall we add a `r.raise_for_status()` to abort on error?
https://requests.readthedocs.io/en/master/user/quickstart/#errors-and-exceptions

DavidSpek replied:
Added

Bobgy reviewed (commented):
/cc @numerology @chensun @Ark-kun 
for review

DavidSpek commented on sdk/python/kfp/_client.py:
```diff
@@ -139,6 +142,11 @@ def __init__(self, host=None, client_id=None, namespace='kubeflow', other_client
     self._experiment_api = kfp_server_api.api.experiment_service_api.ExperimentServiceApi(api_client)
     self._pipelines_api = kfp_server_api.api.pipeline_service_api.PipelineServiceApi(api_client)
     self._upload_api = kfp_server_api.api.PipelineUploadServiceApi(api_client)
+    if self._context_setting['namespace'] == "" and self.get_kfp_healthz().get('multi_user') is True and self._is_ipython() is True:
```
@Bobgy The namespace is currently set by reading `/var/run/secrets/kubernetes.io/serviceaccount/namespace`. I added the check for ipython for situations where th KFP SDK is used outside the cluster in a multi-user environment as it would then not be able to get the current namespace. Another option would be to use the Kubernetes client to get the namespace.

elikatsis replied:
@DavidSpek
But the IPython check isn't doing what you describe. I could be running in an IPython kernel on my laptop.
Reading the `/var/run/secrets/kubernetes.io/serviceaccount/namespace` file is the key. You cannot set the namespace unless reading this file succeeds.

DavidSpek replied:
@elikatsis Your right in that it isn't the correct way to handle this. I think this can best be solved by checking if the `/var/run/secrets/kubernetes.io/serviceaccount/namespace` file exists instead.

elikatsis replied:
@DavidSpek, you could just `try/except` the reading of the file inside that `if`.
I.e., put the `with open(...)...` in a `try/except` block.

DavidSpek replied:
@elikatsis That was also my first intuition but I wasn't sure how to handle the `except`. I now have:
```
except:
    print('Failed to automatically set namespace.')
    pass
```

DavidSpek commented:
@Bobgy The error seen in the e2e-test and sample-test seem to be intermittent. Would you have any idea why and when the healthz API endpoint would give an empty response but not an HTTP error code? Also, it seems like the healthz API has content type `text/plain` rather than `application/json`, is there a reason for this?

DavidSpek commented:
/retest

DavidSpek commented:
@Bobgy The issue still seems to be that the request to the healthz API is returning empty.

Bobgy commented on sdk/python/kfp/_client.py:
```diff
@@ -277,9 +289,32 @@ def set_user_namespace(self, namespace):
+    Returns:
+      response: json formatted response from the healtz endpoint.
+    """
+    healthz_api = 'http://' + self._existing_config.host + '/' + Client.HEALTH_PATH
```
I believe the host should contain schema in it, some hosts may be used with `https` and some may be used with `http`, so I think we should remove `http://` here

Bobgy replied:
in fact, I think your test failure is exactly because of this: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubeflow_pipelines/4638/kubeflow-pipeline-sample-test/1325844927585718272

We had a recent change to use https host for presubmit tests, so it breaks when you add `http://` there

DavidSpek replied:
I was also looking at that, thinking at first `_get_url_prefix` might be an option but that sets the uihost. However, if the empty response was due to an `http` request being sent to an `https` gateway, I would expect an http error that `r.raise_for_status()` would then pick up on.

DavidSpek replied:
I must have written my response just before your second message so I hadn't seen it yet. If you meant that `self._existing_config.host` should already contain `http://` or `https://`, on my cluster it doesn't but that might be outdated as it is kubeflow 1.1. I just pushed a commit changing to `https` to test that theory and hopefully clear things up. If that comes back positive I'll need to figure out a method for determining if it should use `http` or `https`.

DavidSpek replied:
@Bobgy  I think I may have found the problem. In the log it says `KFP API host is 502b4a4db4ab8efb-dot-datalab-vm-staging.googleusercontent.com`, however, due to forcing https the request got an SSL error when it tried to connect to `ml-pipeline.kubeflow.svc.cluster.local`. Is there a way to get the API host other than using `self._existing_config.host`, as it seems the in_cluster host is not returning a response?

DavidSpek replied:
While trying to figure out where I would be able to grab `502b4a4db4ab8efb-dot-datalab-vm-staging.googleusercontent.com` from I noticed that in the `run_sample_test.py` the host is set to `'ml-pipeline.%s.svc.cluster.local:8888' % self._namespace`. Is that expected behavior given the above statement that presumbit tests have been changed to `https`?

DavidSpek replied:
As far as I can tell, requests is handling the redirection from `http` to `https` properly. The final url it lands on is `https://accounts.google.com/ServiceLogin?continue=https%3A%2F%2Fdatalab-staging.cloud.google.com%2F_signin%3Fcontinue%3Dhttps%25253A%25252F%25252F72a2c93edd5b49f8-dot-datalab-vm-staging.googleusercontent.com%25252Fapis%25252Fv1beta1%25252Fhealthz%26endpoint%3D72a2c93edd5b49f8`, which means the above statement that it is using `ml-pipeline.kubeflow.svc.cluster.local` for the API address is not correct. The header that is returned is:
```
{'Content-Type': 'text/html; charset=UTF-8', 'X-Frame-Options': 'DENY', 'x-auto-login': 'realm=com.google&args=continue%3Dhttps%253A%252F%252Fdatalab-staging.cloud.google.com%252F_signin%253Fcontinue%253Dhttps%2525253A%2525252F%2525252F72a2c93edd5b49f8-dot-datalab-vm-staging.googleusercontent.com%2525252Fapis%2525252Fv1beta1%2525252Fhealthz%2526endpoint%253D72a2c93edd5b49f8', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'Pragma': 'no-cache', 'Expires': 'Mon, 01 Jan 1990 00:00:00 GMT', 'Date': 'Wed, 11 Nov 2020 14:49:06 GMT', 'Transfer-Encoding': 'chunked', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'Content-Security-Policy': "script-src 'nonce-RHIL2rIpv6K3J5D6qGt4lQ' 'unsafe-inline' 'unsafe-eval';object-src 'none';base-uri 'self';report-uri /cspreport", 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '1; mode=block', 'Server': 'GSE', 'Set-Cookie': '__Host-GAPS=1:2RJhKy8dLzlqLoYcVdWNgDaydS-zkA:HWtOcjrS-7olt-5P;Path=/;Expires=Fri, 11-Nov-2022 14:49:06 GMT;Secure;HttpOnly;Priority=HIGH'}
```

The response content seems to be a google login page or something similar. 

@Bobgy Do you know the type of restrictions there are on accessing the API endpoint? Does a specific header or cookie need to be passed along for example?

Bobgy commented on sdk/python/kfp/_client.py:
```diff
@@ -277,9 +289,32 @@ def set_user_namespace(self, namespace):
+    healthz_api = 'http://' + self._existing_config.host + '/' + Client.HEALTH_PATH
+    count = 0
+    response = None
+    max_attempts = 60
```
Do we really need to retry 60 times? If so, shall we rename the function to `wait_kfp_healthz`?

DavidSpek replied:
This was mainly just to try and brute-force getting a response to see if retries would help. If this indeed seems necessary I think the retry should be handled when trying to set the namespace rather than here.

Bobgy commented on sdk/python/kfp/_client.py:
```diff
@@ -277,9 +289,32 @@ def set_user_namespace(self, namespace):
+        return response
+      except ValueError:
+        print("Exception; retrying after 5 seconds", ValueError)
+        time.sleep(5)
```
We should raise an timeout exception if the max attempts isn't enough.

numerology replied:
+1

Bobgy reviewed (commented):
Seeing the response to be empty is kind of weird, shall we file a separate PR for it?
Or shall we write a method to wait for healthy API server?

DavidSpek commented:
@Bobgy I also think it is strange, especially seeing as it doesn't seem to be restricted to newly setup clusters (but very rare on my cluster). 
I don't mind trying to solve it by writing a method, but at this point I'm not sure what type of method that should be and where to implement it.

Bobgy commented:
To call the api in the same way as other api calls, you can use this method: https://github.com/kubeflow/pipelines/blob/148dd8dcdb23950a3949060e911caa1177477a2b/backend/api/python_http_client/kfp_server_api/api_client.py#L398

of this instance: https://github.com/kubeflow/pipelines/blob/148dd8dcdb23950a3949060e911caa1177477a2b/sdk/python/kfp/_client.py#L133-L135

numerology commented on sdk/python/kfp/_client.py:
```diff
@@ -32,6 +33,8 @@
 from kfp.compiler._k8s_helper import sanitize_k8s_name
 
 from kfp._auth import get_auth_token, get_gcp_access_token
+from ssl import SSLError
```
nit: prefer import module instead of classes

numerology commented on sdk/python/kfp/_client.py:
```diff
@@ -139,6 +144,16 @@ def __init__(self, host=None, client_id=None, namespace='kubeflow', other_client
     self._pipelines_api = kfp_server_api.api.pipeline_service_api.PipelineServiceApi(api_client)
     self._upload_api = kfp_server_api.api.PipelineUploadServiceApi(api_client)
+    self._api_client = api_client
+    if self._context_setting['namespace'] == '' and self.get_kfp_healthz().get('multi_user') is True:
```
Should it be the following, just to be safer?
```suggestion
    if not self._context_setting['namespace'] and self.get_kfp_healthz().get('multi_user') is True:
```

numerology commented on sdk/python/kfp/_client.py:
```diff
@@ -139,6 +144,16 @@ def __init__(self, host=None, client_id=None, namespace='kubeflow', other_client
     self._upload_api = kfp_server_api.api.PipelineUploadServiceApi(api_client)
+    self._api_client = api_client
+    if self._context_setting['namespace'] == '' and self.get_kfp_healthz().get('multi_user') is True:
+      NAMESPACE_PATH = '/var/run/secrets/kubernetes.io/serviceaccount/namespace'
```
Can this be a module level constant instead?

numerology commented on sdk/python/kfp/_client.py:
```diff
@@ -139,6 +144,16 @@ def __init__(self, host=None, client_id=None, namespace='kubeflow', other_client
+          self.set_user_namespace(current_namespace)
+      except:
+        print('Failed to automatically set namespace.')
+        pass
```
Is this `pass` required here?

numerology commented on sdk/python/kfp/_client.py:
```diff
@@ -139,6 +144,16 @@ def __init__(self, host=None, client_id=None, namespace='kubeflow', other_client
+          current_namespace = f.read()
+          self.set_user_namespace(current_namespace)
+      except:
+        print('Failed to automatically set namespace.')
```
nit: Consider using proper logging module, e.g., `absl.logging` where users can control the verbosity and the output stream of logging info.

numerology commented on sdk/python/kfp/_client.py:
```diff
@@ -277,9 +292,47 @@ def set_user_namespace(self, namespace):
       json.dump(self._context_setting, f)
 
+  def get_kfp_healthz(self):
+    """Get healthz info of KFP deployment.
```
```suggestion
    """Gets healthz info of KFP deployment.
```

numerology commented on sdk/python/kfp/_client.py:
```diff
@@ -277,9 +292,47 @@ def set_user_namespace(self, namespace):
+        response = json.loads(r.data)
+        return response
+      except JSONDecodeError:
+        print("Exception; No JSON returned. Retrying after 5 seconds", JSONDecodeError)
```
Ditto. If we use `logging` here we can use lazy `%` formatting

numerology reviewed (commented):
Thanks! Some comments from Python perspective

DavidSpek commented:
@numerology Thank you for looking at the Python. I started programming 2 months ago or so, so digging through KFP and figuring out how it works has been difficult and I am totally not up to date on all the best practices.Once I have the code working during the e-e and sample test I will look at cleaning it up. The problem I am having now is that the instance does not seem to be authenticated for the API endpoint and is redirected to a Google sign-in page. I am now using `r = self._api_client.request('GET', 'https://' + healthz_api)` and `r = self._api_client.request('GET', 'http://' + healthz_api)` so that it is using the same instance as the rest of the client and handling both http and https as I don't think redirect has been implemented (and could also potentially cause trouble). The log below shows that the correct URL is being used that is grabbed from `self._existing_config.host` (so it isn't the case that the internal cluster URL is accidentally used). The HTTP status is 200, but it redirects and ends up at a Google sign-in page rather than throwing a 403 it seems. Do you have any idea why I would be facing this issue or how it could potentially be solved?

```
integration-test-7pz7f-2122502961: KFP API host is 38028ea9b355eb02-dot-datalab-vm-staging.googleusercontent.com
integration-test-7pz7f-2122502961: Run the sample tests...
integration-test-7pz7f-2122502961: Exception; No JSON returned. Retrying after 5 seconds <class 'json.decoder.JSONDecodeError'>
integration-test-7pz7f-2122502961: Request was redirected
integration-test-7pz7f-2122502961: 302 https://38028ea9b355eb02-dot-datalab-vm-staging.googleusercontent.com/apis/v1beta1/healthz
integration-test-7pz7f-2122502961: 307 https://datalab-staging.cloud.google.com/_signin?continue=https%253A%252F%252F38028ea9b355eb02-dot-datalab-vm-staging.googleusercontent.com%252Fapis%252Fv1beta1%252Fhealthz&endpoint=38028ea9b355eb02
integration-test-7pz7f-2122502961: Final destination:
integration-test-7pz7f-2122502961: 200 https://accounts.google.com/ServiceLogin?continue=https%3A%2F%2Fdatalab-staging.cloud.google.com%2F_signin%3Fcontinue%3Dhttps%25253A%25252F%25252F38028ea9b355eb02-dot-datalab-vm-staging.googleusercontent.com%25252Fapis%25252Fv1beta1%25252Fhealthz%26endpoint%3D38028ea9b355eb02
integration-test-7pz7f-2122502961: Traceback (most recent call last):
integration-test-7pz7f-2122502961:   File "/python/src/github.com/kubeflow/pipelines/test/sample-test/sample_test_launcher.py", line 260, in <module>
integration-test-7pz7f-2122502961:     main()
integration-test-7pz7f-2122502961:   File "/python/src/github.com/kubeflow/pipelines/test/sample-test/sample_test_launcher.py", line 256, in main
integration-test-7pz7f-2122502961:     'component_test': ComponentTest
integration-test-7pz7f-2122502961:   File "/usr/local/lib/python3.7/dist-packages/fire/core.py", line 138, in Fire
integration-test-7pz7f-2122502961:     component_trace = _Fire(component, args, parsed_flag_args, context, name)
integration-test-7pz7f-2122502961:   File "/usr/local/lib/python3.7/dist-packages/fire/core.py", line 468, in _Fire
integration-test-7pz7f-2122502961:     target=component.__name__)
integration-test-7pz7f-2122502961:   File "/usr/local/lib/python3.7/dist-packages/fire/core.py", line 672, in _CallAndUpdateTrace
integration-test-7pz7f-2122502961:     component = fn(*varargs, **kwargs)
integration-test-7pz7f-2122502961:   File "/python/src/github.com/kubeflow/pipelines/test/sample-test/sample_test_launcher.py", line 196, in run_test
integration-test-7pz7f-2122502961:     pysample_checker.run()
integration-test-7pz7f-2122502961:   File "/python/src/github.com/kubeflow/pipelines/test/sample-test/run_sample_test.py", line 63, in run
integration-test-7pz7f-2122502961:     self._client = Client(host=self._host)
integration-test-7pz7f-2122502961:   File "/usr/local/lib/python3.7/dist-packages/kfp-1.1.1-py3.7.egg/kfp/_client.py", line 148, in __init__
integration-test-7pz7f-2122502961:     if self._context_setting['namespace'] == '' and self.get_kfp_healthz().get('multi_user') is True:
integration-test-7pz7f-2122502961: AttributeError: 'NoneType' object has no attribute 'get'
```

Bobgy commented:
@DavidSpek https://github.com/kubeflow/pipelines/blob/e28ec4d2dfcdaa5a74948c0b5eea16033f56ea37/backend/api/python_http_client/kfp_server_api/api/pipeline_service_api.py#L163-L177

Here's an example

Bobgy commented:
@DavidSpek it seems you are picking in other people's commits in your branch, so that you get the CLA problem. Can you clean up your branch

Bobgy commented:
@DavidSpek I just recall that the presubmit test doesn't use kfp-server-api package in the same commit https://www.github.com/kubeflow/pipelines/tree/master/test%2Fsample-test%2FDockerfile.

But considering PR like yours, we should. I believe you can add an extra install step that installs the local kfp-server-api.

Bobgy commented:
Or you can separate out the PR into two, one for just adding the API and kfp-server-api pypi package.

The second one updates sdk

DavidSpek commented:
@Bobgy @numerology It seems like I have finally succeeded in getting the PR to work. As suggested above, I changed the sample test dockerfile to include installing the kfp_server_api from the current repo. 
In `healthz.proto` I have defined `GetHealthzRequest`, which doesn't serve any purpose. Would it be better to replace that with `google.protobuf.Empty`?
I currently still have it set to retry calling the Healthz API 5 times, I've added a logging message that states the attempts if it fails. Should this retrying remain in the code or should it be taken out?

numerology commented on sdk/python/kfp/_client.py:
```diff
@@ -23,6 +23,7 @@
 import yaml
 import zipfile
 import datetime
+import requests
```
Do you mind updating `sdk/requirements.txt` using `pip-compile` to make sure it reflects the latest dependency closure? Thanks

DavidSpek replied:
Requests is actually not needed anymore. If I run `pip-compile --output-file=requirements.txt requirements.in` two things are removed: `importlib-metadata==1.5.0  # via jsonschema` and `zipp==1.2.0               # via importlib-metadata`. Given that requests is not necessary, it should unchanged. Should I leave it unchanged or the one with those 2 removed?

numerology replied:
Interesting... So I'm fine leave it as is. Thanks

numerology reviewed (commented):
Thanks!

numerology commented:
> I currently still have it set to retry calling the Healthz API 5 times, I've added a logging message that states the attempts if it fails. Should this retrying remain in the code or should it be taken out?

Retrying looks good to me. In addition to that, shall we raise an error when maximal number of attempts is reached?

DavidSpek commented:
@numerology How about this for the retries?
```
def get_kfp_healthz(self):
    """Gets healthz info of KFP deployment.

    Returns:
      response: json formatted response from the healtz endpoint.
    """
    count = 0
    response = None
    max_attempts = 5
    while not response:
      if count < max_attempts:
        raise TimeoutError('API call timeout')
      count += 1
      try:
        response = self._healthz_api.get_healthz()
        return response
      except:
        logging.info('Failed to get healthz info attempt {}.'.format(max_attempts))
        time.sleep(5)
```

numerology commented:
/lgtm
for the SDK and Python part.

Good job! I'll leave the final approval to @Bobgy

DavidSpek commented:
/test kubeflow-pipeline-frontend-test

NikeNano commented on go.mod:
```diff
@@ -31,6 +34,23 @@ require (
 	github.com/mattn/go-sqlite3 v1.9.0
 	github.com/minio/minio-go v6.0.14+incompatible
 	github.com/peterhellberg/duration v0.0.0-20191119133758-ec6baeebcd10
+	github.com/smartystreets/assertions v0.0.0-20180927180507-b2de0cb4f26d // indirect
```
What triggered adding the new dependencies?

Bobgy replied:
I'm guessing it's the bazel build.
@DavidSpek can you try removing updates to this file in one commit and see if presubmit still passes (I believe these should be just temporary dependencies fetched when building the protos?

If it does, we'll need to fix the api generation script, so that it doesn't use bazel any more, but it should be outside of this PR.

DavidSpek replied:
These changes are indeed from running the script. Although it isn't the most reliable, VS code is also lots of these dependencies are not used. I'm removing them now for a new commit to test if things brake.

DavidSpek replied:
Testing was successful with the changes to this file removed, so I believe the api generation script will need some fixing.

NikeNano commented on sdk/python/kfp/_client.py:
```diff
@@ -139,6 +141,14 @@ def __init__(self, host=None, client_id=None, namespace='kubeflow', other_client
+        with open(Client.NAMESPACE_PATH, 'r') as f:
+          current_namespace = f.read()
+          self.set_user_namespace(current_namespace)
+      except:
```
Maybe you could try to only catch the exact type of error that you expect here

Bobgy replied:
also we should log the exception in logging.info

DavidSpek replied:
I agree, however, due to my limited experience with programming and no ability to test this PR fully on my own cluster easily I'm not 100% sure what exception can all be expected. The main one I had in mind is `FileNotFoundError` I believe. So I have added that and added `exc_info=True` to the logging.info line. Hopefully this is correct.

DavidSpek replied:
Testing just passed so I believe this can be considered solved.

Bobgy commented on backend/api/healthz.proto:
```diff
@@ -0,0 +1,85 @@
+  }
+
+message GetHealthzResponse {
+  // The commit sha.
+  string commit_sha = 1;
+
+  // The tag name
+  string tag_name = 2;
```
Shall we remove unused fields first?
Because when accessing healthz api from frontend proxy, the two fields will be renamed to `apiServerCommitSha`. https://github.com/Bobgy/pipelines/blob/6c326ad8eab30387fc18f6b5e8b09bea6d6c52e5/frontend/server/handlers/healthz.ts#L68-L87

We need a horizontal alignment on the best way forward.

Bobgy replied:
To be exact, immediate action items:
1. remove the two fields
2. regenerate python client

DavidSpek replied:
I've just pushed the changes. If I understand correctly `multi_user = 3` is still correct and it should not be changed to `multi_user = 1`.

DavidSpek replied:
Tests completed successfully so hopefully it is alright.

Bobgy commented on sdk/python/kfp/_client.py:
```diff
@@ -115,6 +115,8 @@ class Client(object):
   IN_CLUSTER_DNS_NAME = 'ml-pipeline.{}.svc.cluster.local:8888'
   KUBE_PROXY_PATH = 'api/v1/namespaces/{}/services/ml-pipeline:http/proxy/'
 
+  NAMESPACE_PATH = '/var/run/secrets/kubernetes.io/serviceaccount/namespace'
```
Can you add a comment linking to documentation saying this file is auto populated in kubernetes pods?

Bobgy reviewed (commented):
Some cleanup comments, thanks a lot again @davidspek!

DavidSpek commented on backend/api/healthz.proto:
```diff
@@ -0,0 +1,85 @@
+  }
+}
+
+message GetHealthzRequest {
```
At one point I had `google.protobuf.Empty` there but decided against it and thus never pushed it to run testing. I'll use that now and see how it goes.

DavidSpek replied:
Testing passed with empty request so marking this as solved

Bobgy commented:
/lgtm
/approve

Thank you so much!
Really appreciated the perseverance!

Ark-kun commented on sdk/python/kfp/_client.py:
```diff
@@ -139,6 +144,14 @@ def __init__(self, host=None, client_id=None, namespace='kubeflow', other_client
     self._experiment_api = kfp_server_api.api.experiment_service_api.ExperimentServiceApi(api_client)
     self._pipelines_api = kfp_server_api.api.pipeline_service_api.PipelineServiceApi(api_client)
     self._upload_api = kfp_server_api.api.PipelineUploadServiceApi(api_client)
+    self._healthz_api = kfp_server_api.api.healthz_service_api.HealthzServiceApi(api_client)
```
Are you sure this line will work with the latest stable version of `kfp_server_api`?
/cc @Bobgy @numerology

DavidSpek replied:
Are people installing mismatching versions of `kfp_server_api` and the kfp client? I think that might be causing the error. If that is the case, maybe a solution would be to check the version of `kfp_server_api` before running this line?

Ark-kun replied:
When installing packages, python does not upgrade dependencies if they are already installed and satisfy the requirements. So, unless the KFP package demands a newer version of `kfp_server_api`, it won't be updated.

DavidSpek replied:
I thought the requirement was changed as it was already the recommendation.

Ark-kun commented on test/sample-test/Dockerfile:
```diff
@@ -11,6 +11,10 @@ RUN wget https://bootstrap.pypa.io/get-pip.py && python3 get-pip.py
 RUN pip3 install -r /python/src/github.com/kubeflow/pipelines/test/sample-test/requirements.txt
 
+# Install KFP server API from commit.
+COPY ./backend/api/python_http_client /python_http_client
```
I'm not sure this approach is correct. This is not how the end users are installing the kfp package.
The changes in this PR could still have been submitted without these lines. They'd just have to be split into two granular PRs which is usually a good idea.

DavidSpek replied:
Without this line the e-2-e testing would fail as there would be a mismatch between the kfp client and kfp-api-server.

Ark-kun replied:
That's the goal of the e2e tests - to detect when things are going to fail for our users.
Here we're essentially disabling part of the testing by performing manual installation.

A better approach might have been to make atomic PRs:

1. Add new API to backend
2. Expose API in the generated clients
3. Use the clients in the SDKs

DavidSpek replied:
This was discussed here: https://github.com/kubeflow/pipelines/pull/4638#issuecomment-729620679. Seeing as the kfp-server-api and the SDK depend on each other, when installing from master they should both be installed from master.

Ark-kun commented:
Perhaps we should have split this PR in two: Backend API and kfp.Client.
This would have allowed a smooth gradual transition from one working state to another.
More focused and granular PRs are also easier to review.
